### PR TITLE
fix: avoid excessive memory allocations

### PR DESCRIPTION
**Pull request by harnash**
This PR changes internal storage for HIBP API call results to use ristretto cache instead of plain map to avoid excessive memory allocations when handling substantial load (each checked hash is stored indefinitely in memory).

## Related issue(s)
* https://github.com/ory/kratos/issues/2354

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

## Further Comments
This patch was running for a few days on our cluster and it looks like memory usage is not increasing and stabilised around 70MB which is great :-)

### CODE REVIEW

File: selfservice/strategy/password/validator.go [hunk 1]
```diff
@@ -62,12 +62,21 @@ type validatorDependencies interface {
 	config.Provider
 }
 
-func NewDefaultPasswordValidatorStrategy(reg validatorDependencies) *DefaultPasswordValidator {
+func NewDefaultPasswordValidatorStrategy(reg validatorDependencies) (*DefaultPasswordValidator, error) {
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 10 * 10000,
+		MaxCost:     60 * 10000, // BCrypt hash size is 60 bytes
+		BufferItems: 64,
```
Comment (aeneasr) [inline selfservice/strategy/password/validator.go:69]: The cost in ristretto is set is kinda weird. For example, it is probably required to ignore the internal cost and use a custom cost function:

https://github.com/dgraph-io/ristretto/blob/8e850b710d6df0383c375ec6a7beae4ce48fc8d5/cache.go#L137

**Review by aeneasr — CHANGES_REQUESTED** [review]
Awesome, thank you for your contribution! This looks pretty good and I have some ideas how to improve it further :)

File: selfservice/strategy/password/validator.go (hunk 1 above)
Comment (harnash) [reply selfservice/strategy/password/validator.go]: Alright. I've added `ignoreInternalCost=true` but I'm not sure if custom cost function is needed at this point.

File: selfservice/strategy/password/validator.go [hunk 2]
```diff
@@ -109,9 +120,8 @@ func (s *DefaultPasswordValidator) fetch(hpw []byte, apiDNSName string) error {
 		return errors.Wrapf(ErrUnexpectedStatusCode, "%d", res.StatusCode)
 	}
 
-	s.Lock()
-	s.hashes[b20(hpw)] = 0
-	s.Unlock()
+	s.hashes.SetWithTTL(b20(hpw), 0, 1, hashCacheItemTTL)
+	s.hashes.Wait()
```
Comment (aeneasr) [inline selfservice/strategy/password/validator.go:124]: Is the wait intentional? I think we could do without!

Comment (harnash) [reply selfservice/strategy/password/validator.go]: Fix in 3c5caa669

**Pull request merged**